assignments/serde-lifetimes.adoc

```diff
@@ -0,0 +1,148 @@
+= Exercise: Copy-on-Write Data with Serde
+:source-language: rust
+
+In this exercise you will learn
+
+* How to serialize and deserialize data with Serde crate.
+* When to use types that own their data or only refer to externally owned data
+* How to use `std::borrow::Cow` type to minimize extra memory allocations
+
+== Getting Started
+
+Create a new binary project for this exercise. Add the following dependencies to your `Cargo.toml` file:
+
+[source,toml]
+----
+[dependencies]
+serde = { version = "1", features = ["derive"] }
+serde_json = "1"
+----
+
+Use this template:
+[source,rust]
+----
+// pretend that we call an API and get a JSON String back
+fn fetch_data() -> String {
+    String::from(
+        r#"
+            {
+                "id": 1,
+                "title": "Hello, Rust"
+            }
+        "#,
+    )
+}
+
+#[derive(Debug)]
+struct BlogPost {
+    id: u32,
+
+    title: String,
+}
+
+fn main() -> Result<(), serde_json::Error> {
+    let post: BlogPost = {
+        let data = fetch_data();
```

Comment:
@listochkin I am confused here. wouldn't this be dropped at the end of the block?

Reply:
That's the point. Step 2: Deserialization using `str` has the following text:

> Use lifetime annotations and change the code inside main if necessary.
>
> Q: If you made a change, why?

The idea is to let the participants stumble on your structure potentially outliving the underlying data storage, and that switch to reference-basedstructures can cause potential re-design of your code.